docs/design/api-versions.md

```diff
@@ -32,8 +40,169 @@ For each API spec, generated `AssignPropertiesTo*()` method  will read the value
 
 For each Storage spec, the generated `AssignPropertiesTo*()` and `AssignPropertiesFrom*()` methods will copy the `OriginalVersion` property between versions, preserving the original value.
 
-### API Recovery
-
-Into each storage resource variant, we'll inject a function `OriginalGVK(): GroupVersionKind` which will use `OriginalVersion` to create the GVK required.
+## New conversion interfaces
 
-SequenceDiagram TBC.
+We don't want the overhead of doing a complete conversion of each resource if we're only interested in conversion of the associated `Spec` or `Status`. For example, when submitting a request to ARM we need only do a `Spec` conversion; similiarly, when processing a response from ARM, we need only do a `Status` conversion.
+
+We already have the required infrastructure in the form of the `AssignProperties*()` methods, making it simple to implement these new interfaces.
+
+For conversion of `Spec` types:
+
+``` go
+type ConvertibleSpec interface {
+    // ConvertFromSpec populates the current spec from the passed one
+    ConvertFromSpec(source ConvertibleSpec) error
+    
+    // ConvertToSpec populates the passed spec from the current one
+    ConvertToSpec(destination ConvertibleSpec) error
+}
+```
+
+As we have done with implementations of the traditional [`Convertible`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/conversion#Convertible) interface, these will be implemented on Specs that are further from our hub, and will chain towards that hub.
+
+For conversion of `Status` types, a similar approach:
+
+``` go
+type ConvertibleStatus interface {
+    // ConvertFromStatus populates the current status from the passed one
+    ConvertFromStatus(source ConvertibleStatus) error
+    
+    // ConvertToStatus populates the passed status from the current one
+    ConvertToStatus(destination ConvertibleStatus) error
+}
+```
+
+## Reconciler integration
+
+When the reconciler is working with ARM, we'll use the `OriginalVersion` property on the storage spec to pivot from the current version to the original version.
+
+To keep the reconciler a simple as possible, we'll handle this with generated code hidden behind a simple interface:
+
+``` go
+type GeneratedResource interface {
```

Comment:
minor: I think a name other than `GeneratedResource` would make sense here? We already have an existing interface which is `KubernetesResource`, documented as:

```
// KubernetesResource is an Azure resource. This interface contains the common set of
// methods that apply to all ASO resources.
type KubernetesResource interface {
	// Owner returns the ResourceReference of the owner, or nil if there is no owner
	Owner() *ResourceReference

	// TODO: I think we need this?
	// KnownOwner() *KnownResourceReference

	// AzureName returns the Azure name of the resource
	AzureName() string

	// Some types, but not all, have a corresponding:
	// 	SetAzureName(name string)
	// They do not if the name must be a fixed value (like 'default').

	// TODO: GetAPIVersion here?

	// TODO: I think we need this
	// SetStatus(status interface{})
}
```

Probably makes sense to just expand that? Certainly this interface and it seem to have a lot in common, and there are some TODOs on `KubernetesResource` that look a lot like what you're proposing here.

Reply:
Or if we want a separate interface we could embed that interface into this interface.

Reply:
Yes, we need a better name than `GeneratedResource`, but that's the best I could come up with.

Not sure it makes sense to put onto `KubernetesResource`, but _maybe_. 

I'm not familiar with the purpose of `KubernetesResource` - if it fits, then sure. But, given interfaces represent abstractions, if the new methods don't fit the abstraction, we should use a new interface.

Reply:
What about `VersionAwareResource` as that's the key functionality hidden behind the curtain - it will automatically do the right thing wrt versions.

Reply:
> But, given interfaces represent abstractions, if the new methods don't fit the abstraction, we should use a new interface.

The original intent of `KubernetesResource` was to define an interface which must be implemented for a resource to be reconciled by the generic reconciler. If it ends up being composed of other interfaces I think that's fine, but I think we should have a single top level interface that is the canonical "This is what you must implement for generic_reconciler/azure_deployment_reconciler to work. This also eases the amount of casting done in the controller because it just has to cast once (to the top level interface) which has everything it's going to need.

Reply:
Ah sorry, actually the interface I was thinking of was:
```
type MetaObject interface {
	runtime.Object
	metav1.Object
	KubernetesResource
}
```
`genruntime.MetaObject` - so yeah we should decide what makes sense in `KubernetesResource` and what makes sense in another interface, but we should put whatever that other interface is into `MetaObject` alongside `KubernetesResource`

Reply:
happy to chat about this in person if you wanna discuss it more

Reply:
On the naming: `VersionAwareResource` feels kinda weird because I'm not sure I would call the API types "version aware" - they only really know their own version, yet they would be implementing this interface. I think that interface name fits well if we were only implementing it on storage types.

Maybe?
```go
// TODO: ARMConverterSupplier is a real awkward name, could be something else.
type ARMConverterSupplier interface {
        GetARMSpecConverter(scheme runtime.Scheme) (genruntime.ARMTransformer, error)
        NewARMStatusConverter(scheme runtime.Scheme) (genruntime.FromARMConverter, error)
}

type KubernetesResource interface {
        ARMConverterSupplier

	// Owner returns the ResourceReference of the owner, or nil if there is no owner
	Owner() *ResourceReference

	// AzureName returns the Azure name of the resource
	AzureName() string

	SetStatus(status interface{}) error
}
```

I pulled `SetStatus` out into the other interface because it felt like it was really pretty generic? In my head at least `KuberentesResource` would really end up having these 4 methods:
```go
GetStatus() interface{}
SetStatus(status interface{}) error
GetSpec() interface{}
SetSpec(spec interface{}) error
```

Although I am not sure we'll need the `SetSpec` one, or the `GetStatus` one because of the directionality of the operator.

A random aside: It's awkward we use both `Transformer` and `Converter` suffixes in the `FromARMConverter`/`ARMTransformer` interfaces. We should probably change `ARMTransformer` to `ARMConverter`? Then it's built of two embedded interfaces `FromARMConverter` and `ToARMConverter`?

Reply:
As discussed in our design review, I'll put the required members directly on `KubernetesResource`, rather than defining a new resource interface. I've updated the spec to reflect this.